src/test/fio/ceph-librgw.fio

```diff
@@ -0,0 +1,32 @@
+[global]
+ioengine=external:/home/mbenjamin/ceph-cp/build/lib/libfio_librgw.so
+rw=randwrite
+name=fiotest
+direct=0
+access_key=${AWS_ACCESS_KEY_ID}
+secret_key=${AWS_SECRET_ACCESS_KEY}
+userid=testuser
+ceph_cluster=ceph
+ceph_conf=/home/mbenjamin/ceph-cp/build/ceph.conf
+#in current impl, there is only one, global bucket
+bucket_name=fiotest
+thread=1
```

Comment:
please try to add `max-jobs=4` (complementing the `numjobs=4` below)

Reply:
thanks, marc--that, plus nr_files=<larger value> make the report look more sane, but I don't feel like the parallelism == min(nr_files, numjobs, max_jobs)

Reply:
@mattbenjamin setting up a test environment, will run a load test and update

Reply:
it seems that fio is overriding the `--thread=0` (on some engines) and creates the pthreads instead of fork() ing - which is why we dont see the performance scale with increasing of `--numjobs=64` and `--max-jobs=64`

diagnostics:
```
sudo ./bin/fio --ioengine=external:/mnt/osd--host/src-git/ceph--upstream--master/build/lib/libfio_librgw.so --name=fiotest --access_key=b2345678901234567890 --secret_key=b234567890123456789012345678901234567890 --userid=cosbench --ceph_cluster=ceph --ceph_conf=/mnt/osd--host/src-git/ceph--upstream--master/build/ceph.conf --bucket_name=fiobkt --nr_files=4000 --bs=4k --size=10G --rw=randwrite --randrepeat=0 --offset=0 --time_based=1 --runtime=30s --numjobs=64 --max-jobs=64 --thread=0 --direct=0 --disable_lat=1 --disable_clat=1 --disable_bw=1 --clat_percentiles=0 --group_report=1
fiotest: (g=0): rw=randwrite, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=librgw, iodepth=1
...
fio-3.15
Starting 64 processes
Jobs: 1 (f=1861): [w(1),_(63)][19.4%][w=1485KiB/s][w=371 IOPS][eta 00m:25s]
fiotest: (groupid=0, jobs=1): err= 0: pid=148744: Mon Jan 11 19:09:06 2021
  write: IOPS=369, BW=1478KiB/s (1514kB/s)(43.3MiB/30001msec); 0 zone resets
```
note: `--numjobs=64 --max-jobs=64 --thread=0` --> `Starting 64 processes` --> `Jobs: 1 ...`
![image](https://user-images.githubusercontent.com/31659604/104216637-fc283400-5442-11eb-8bec-46e0c1aa1425.png)

bisecting: `--ioengine=null` works correctly:
```
fio --ioengine=null --max-jobs=64 --numjobs=64 --name=maxjobs --size=10M --time_based=1 --runtime=30s --group_report=1
Starting 64 processes
Jobs: 64 (f=64): [R(64)][12.9%][r=212GiB/s][r=55.6M IOPS][eta 00m:27s]
maxjobs: (groupid=0, jobs=64): err= 0: pid=168133: Mon Jan 11 19:16:20 2021
```
note the: `Jobs: 64`

Reply:
to scale the performance for testing the envelope it's possible to launch multiple instances of fio manually
```
for i in {1..4}; do \
sudo ./bin/fio --ioengine=external:/mnt/osd--host/src-git/ceph--upstream--master/build/lib/libfio_librgw.so --name=fiotest --access_key=b2345678901234567890 --secret_key=b234567890123456789012345678901234567890 --userid=cosbench --ceph_cluster=ceph --ceph_conf=/mnt/osd--host/src-git/ceph--upstream--master/build/ceph.conf --bucket_name=fiobkt --nr_files=4000 --bs=4k --size=10G --rw=randwrite --randrepeat=0 --offset=0 --time_based=1 --runtime=30s --numjobs=8 --max-jobs=8 --thread=0 --direct=0 --disable_lat=1 --disable_clat=1 --disable_bw=1 --clat_percentiles=0 --group_report=1 | grep "IOPS=" & ;\
done
  write: IOPS=299, BW=1197KiB/s (1226kB/s)(35.1MiB/30001msec); 0 zone resets
  write: IOPS=304, BW=1216KiB/s (1246kB/s)(35.6MiB/30002msec); 0 zone resets
  write: IOPS=300, BW=1203KiB/s (1232kB/s)(35.3MiB/30003msec); 0 zone resets
  write: IOPS=301, BW=1204KiB/s (1233kB/s)(35.3MiB/30003msec); 0 zone resets
# 300 * 4 = 1200 IOPS


for i in {1..8}; do \
sudo ./bin/fio --ioengine=external:/mnt/osd--host/src-git/ceph--upstream--master/build/lib/libfio_librgw.so --name=fiotest --access_key=b2345678901234567890 --secret_key=b234567890123456789012345678901234567890 --userid=cosbench --ceph_cluster=ceph --ceph_conf=/mnt/osd--host/src-git/ceph--upstream--master/build/ceph.conf --bucket_name=fiobkt --nr_files=4000 --bs=4k --size=10G --rw=randwrite --randrepeat=0 --offset=0 --time_based=1 --runtime=30s --numjobs=8 --max-jobs=8 --thread=0 --direct=0 --disable_lat=1 --disable_clat=1 --disable_bw=1 --clat_percentiles=0 --group_report=1 | grep "IOPS=" & ;\
done
  write: IOPS=207, BW=829KiB/s (849kB/s)(24.3MiB/30001msec); 0 zone resets
  write: IOPS=207, BW=829KiB/s (849kB/s)(24.3MiB/30003msec); 0 zone resets
  write: IOPS=203, BW=815KiB/s (835kB/s)(23.9MiB/30001msec); 0 zone resets
  write: IOPS=205, BW=821KiB/s (841kB/s)(24.1MiB/30002msec); 0 zone resets
  write: IOPS=204, BW=820KiB/s (840kB/s)(24.0MiB/30001msec); 0 zone resets
  write: IOPS=203, BW=814KiB/s (834kB/s)(23.9MiB/30003msec); 0 zone resets
  write: IOPS=203, BW=814KiB/s (833kB/s)(23.8MiB/30002msec); 0 zone resets
  write: IOPS=204, BW=817KiB/s (837kB/s)(23.9MiB/30001msec); 0 zone resets
# 200 * 8 = 1600 IOPS
```
In the latter test, the performance is capped by 100% iostat io utilization on the test system drives